sound/soc/sof/sof-client-probes-ipc4.c

```diff
@@ -0,0 +1,289 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// Copyright(c) 2019-2022 Intel Corporation. All rights reserved.
+//
+// Author: Jyri Sarha <jyri.sarha@intel.com>
+//
+
+#include <sound/soc.h>
+#include <sound/sof/ipc4/header.h>
+#include <uapi/sound/sof/header.h>
+#include "sof-priv.h"
+#include "ipc4-priv.h"
+#include "sof-client.h"
+#include "sof-client-probes.h"
+#include "ipc4-topology.h"
+
+enum sof_ipc4_dma_type {
+	SOF_IPC4_DMA_HDA_HOST_OUTPUT = 0,
+	SOF_IPC4_DMA_HDA_HOST_INPUT = 1,
+	SOF_IPC4_DMA_HDA_LINK_OUTPUT = 8,
+	SOF_IPC4_DMA_HDA_LINK_INPUT = 9,
+	SOF_IPC4_DMA_DMIC_LINK_INPUT = 11,
+	SOF_IPC4_DMA_I2S_LINK_OUTPUT = 12,
+	SOF_IPC4_DMA_I2S_LINK_INPUT = 13,
+};
+
+union sof_ipc4_connector_node_id {
+	u32 val;
+	struct {
+		u32 vindex:8;
+		u32 dma_type:5;
+		u32 rsvd:19;
+	};
+} __packed;
+
+enum sof_ipc4_probe_runtime_param {
+	SOF_IPC4_PROBE_INJECTION_DMA = 1,
+	SOF_IPC4_PROBE_INJECTION_DMA_DETACH,
+	SOF_IPC4_PROBE_POINTS,
+	SOF_IPC4_PROBE_POINTS_DISCONNECT,
+};
+
+/* For getting probe dmas */
+struct sof_ipc4_probe_dma {
+	union sof_ipc4_connector_node_id node_id;
+	unsigned int dma_buffer_size;
+} __packed;
+
+struct sof_ipc4_probe_gtw_cfg {
+	union sof_ipc4_connector_node_id node_id;
+	u32 dma_buffer_size;
+} __packed;
+
+struct sof_ipc4_probe_cfg {
+	struct sof_ipc4_base_module_cfg base;
+	struct sof_ipc4_probe_gtw_cfg gtw_cfg;
+} __packed;
+
+enum sof_ipc4_probe_type {
+	SOF_IPC4_PROBE_TYPE_INPUT = 0,
+	SOF_IPC4_PROBE_TYPE_OUTPUT,
+	SOF_IPC4_PROBE_TYPE_INTERNAL
+};
+
+union sof_ipc4_probe_point_id {
+	u32 full_id;
+	struct {
+		u32 module_id:16;
+		u32 instance_id:8;
+		u32 type:2;
+		u32 index:6;
+	} fields;
+} __packed;
+
+struct sof_ipc4_probe_point {
+	union sof_ipc4_probe_point_id buffer_id;
+	uint32_t purpose;
+	uint32_t stream_tag;
+} __packed;
+
+#define INVALID_PIPELINE_ID      0xFF
+
+static struct sof_man4_module *sof_ipc4_probe_get_module_info(struct sof_client_dev *cdev)
+{
+	struct sof_probes_priv *priv = cdev->data;
+	struct snd_sof_dev *sdev = cdev->sdev;
+	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+	struct sof_ipc4_fw_module *fw_modules = ipc4_data->fw_modules;
+	static const guid_t probe_uuid =
+		GUID_INIT(0x7CAD0808, 0xAB10, 0xCD23,
+			  0xEF, 0x45, 0x12, 0xAB, 0x34, 0xCD, 0x56, 0xEF);
+	int i;
+
+	if (!priv->ipc_priv) {
+		if (!fw_modules) {
+			dev_err(sdev->dev, "no fw_module information\n");
+			return NULL;
+		}
+
+		/* set module info */
+		for (i = 0; i < ipc4_data->num_fw_modules; i++) {
+			if (guid_equal(&probe_uuid,
+				       &fw_modules[i].man4_module_entry.uuid)) {
+				priv->ipc_priv = &fw_modules[i].man4_module_entry;
+				break;
+			}
+		}
+	        if (!priv->ipc_priv)
+			dev_err(sdev->dev, "%s: no matching uuid found", __func__);
+	}
+
+	return (struct sof_man4_module *) priv->ipc_priv;
```

Comment:
I think you will need to rework this after https://github.com/thesofproject/linux/pull/3826 is merged and have a sof-client wrapper to get a module by uuid..